htdocs/core/modules/barcode/mod_barcode_product_standard.php

```diff
@@ -171,7 +172,22 @@ public function getNextValue($objproduct = null, $type = '')
 		$now = dol_now();
 
 		$numFinal = get_next_value($db, $mask, 'product', $field, $where, '', $now);
-
+		//if EAN13 calculate and substitute the last 13th character (* or ?) used in the mask by the EAN13 key
+		$type=$conf->global->PRODUIT_DEFAULT_BARCODE_TYPE; //get barcode type configuration for products
```

Comment:
The $type is already defined and passed to the function. It may be different than PRODUIT_DEFAULT_BARCODE_TYPE for example when function is called to getNextValue for something else than an EAN13. Why adding this line ?

Reply:
Hi Eldy,
In fact I also thought that $type was passed to the function but I put a tracer on it and it only returned 0 while I was waiting for EAN information. That's why I decided to test the EAN configuration for the products because that's what automatic generation is used for.

Reply:
Hello,
I am very motivated to make this EAN13 barcode generation function definitively operational, on the one hand because it is indispensable to me and my users, but on the other hand because it's a problem for a significant number of people, some of whom have posted on the forums, particularly in French, for "some time": [example 1](https://www.dolibarr.fr/forum/t/generateur-code-barre/24076/6), [example 2](https://www.dolibarr.fr/forum/t/probleme-de-code-barre/27070), [example 3](https://www.dolibarr.fr/forum/t/generation-automatique-codes-barres-des-produits/20862/3), [example 4](https://www.dolibarr.fr/forum/t/utiliser-code-barre-lors-de-creation-de-proposition-commerciale/32597/3).
Thank you for telling me how I can help further in this direction.

Reply:
$type is code field in c_barcode_type (EAN13) coming from commonobject ->fetch_barcode(), not sure why you have null.
Shouldn't be $mask checked for * or ? and not $numFinal?

Reply:
Hi Fappels,
- may be I get NULL because of an upstream bug but I didn't identify it, so to bypass I found an other (but not perfect) solution.
- $mask is fixed and contains (for EAN13) for example 020{000000000}? , $numFinal is the calculated next value from the $mask, following this example 020000000095? , by chance, reading the code and experimenting it, I found that the function get_next_value in the preceding line:
`$numFinal = get_next_value($db, $mask, 'product', $field, $where, '', $now);`
if the last character is only * or ? (fonctionning as jokers), can calculate the next value on the first 12 digits of the mask, leaving the last character (* or ?) unchanged. So you have just to test it and replace it by the EAN13 key (which calculation function barcode_gen_ean_sum already exists on /core/lib/barcode.lib.php.
That's what my code does.

Reply:
@fappels, @eldy 
it seems that public function fetch_barcode() is not called for the barcode automatic calculation even to give the next value on the codebarre module settings, nor to give the next value on the product codebarre field. 
It is only called to show the codebarre type on the product type field.
So it could explain why $type is null/empty on public function getNextValue of mod_barcode_product_standard.php if it is never passed...
For the product I can easily improve this in htdocs/product/card.php
I will propose you soon an improved code, replacing
`$tmpcode = $modBarCodeProduct->getNextValue($object, $type);`
by
`$tmpcode = $modBarCodeProduct->getNextValue($object, $fk_barcode_type);`
and testing it consequently.

Reply:
> Hi Eldy,
> In fact I also thought that $type was passed to the function but I put a tracer on it and it only returned null while I was waiting for EAN information (it's true setting a product barcode and showing the next generated barcode in the module config). That's why I decided to test the EAN configuration for the products because that's what automatic generation is used for.

So if it is null, there is a problem with caller. You can add a workaround to fix this but you must make
$type=$conf->global->PRODUIT_DEFAULT_BARCODE_TYPE; 
only if it is not already defined.